File: backend-nodejs/src/controllers/user.controller.ts
import { Request, Response } from 'express';
import { validationResult } from 'express-validator';
import { UserService } from '../services/user.service';
import { generateToken, getTokenExpirationTime } from '../utils/jwt.utils';
import { 
  CreateUserRequest, 
  LoginRequest, 
  UpdateUserRequest,
  JwtResponse, 
  HealthResponse,
  ApiResponse
} from '../types/user.types';

/**
 * 사용자 컨트롤러 클래스
 * REST API 엔드포인트 처리
 */
export class UserController {
  private userService: UserService;

  constructor() {
    this.userService = new UserService();
  }

  /**
   * 헬스체크 엔드포인트
   */
  healthCheck = async (req: Request, res: Response): Promise<void> => {
    const healthData: HealthResponse = {
      status: 'healthy',
      timestamp: Math.floor(Date.now() / 1000),
      app_name: 'Sample Intranet Backend Node.js'
    };
    res.json(healthData);
  };

  /**
   * 회원가입
   */
  register = async (req: Request, res: Response): Promise<void> => {
    try {
      // 입력 검증 확인
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        res.status(400).json({
          error: '유효성 검사 실패',
          detail: errors.array().map(err => err.msg).join(', ')
        });
        return;
      }

      const userData: CreateUserRequest = req.body;
      const user = await this.userService.createUser(userData);

      res.status(201).json(user);
    } catch (error) {
      res.status(400).json({
        error: '회원가입 실패',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };

  /**
   * 로그인
   */
  login = async (req: Request, res: Response): Promise<void> => {
    try {
      // 입력 검증 확인
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        res.status(400).json({
          error: '유효성 검사 실패',
          detail: errors.array().map(err => err.msg).join(', ')
        });
        return;
      }

      const { email, password }: LoginRequest = req.body;
      const user = await this.userService.authenticateUser(email, password);

      if (!user) {
        res.status(401).json({
          error: '로그인 실패',
          detail: '이메일 또는 비밀번호가 올바르지 않습니다'
        });
        return;
      }

      // JWT 토큰 생성
      const token = generateToken({
        userId: user.id,
        email: user.email,
        role: user.role
      });

      const response: JwtResponse = {
        accessToken: token,
        tokenType: 'bearer',
        expiresIn: `${getTokenExpirationTime()}s`,
        user
      };

      res.json(response);
    } catch (error) {
      res.status(500).json({
        error: '로그인 처리 중 오류 발생',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };

  /**
   * 사용자 목록 조회
   */
  getUsers = async (req: Request, res: Response): Promise<void> => {
    try {
      const skip = parseInt(req.query.skip as string) || 0;
      const limit = parseInt(req.query.limit as string) || 100;
      const name = req.query.name as string;

      const users = await this.userService.getAllUsers({ skip, limit, name });
      res.json(users);
    } catch (error) {
      res.status(500).json({
        error: '사용자 목록 조회 실패',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };

  /**
   * 특정 사용자 조회
   */
  getUser = async (req: Request, res: Response): Promise<void> => {
    try {
      const id = parseInt(req.params.id);
      if (isNaN(id)) {
        res.status(400).json({
          error: '잘못된 사용자 ID',
          detail: '유효한 숫자 ID를 입력해주세요'
        });
        return;
      }

      const user = await this.userService.getUserById(id);
      if (!user) {
        res.status(404).json({
          error: '사용자를 찾을 수 없습니다',
          detail: `ID ${id}에 해당하는 사용자가 없습니다`
        });
        return;
      }

      res.json(user);
    } catch (error) {
      res.status(500).json({
        error: '사용자 조회 실패',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };

  /**
   * 사용자 정보 수정
   */
  updateUser = async (req: Request, res: Response): Promise<void> => {
    try {
      const id = parseInt(req.params.id);
      if (isNaN(id)) {
        res.status(400).json({
          error: '잘못된 사용자 ID',
          detail: '유효한 숫자 ID를 입력해주세요'
        });
        return;
      }

      const updateData: UpdateUserRequest = req.body;
      const updatedUser = await this.userService.updateUser(id, updateData);

      res.json(updatedUser);
    } catch (error) {
      res.status(400).json({
        error: '사용자 정보 수정 실패',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };

  /**
   * 사용자 삭제 (관리자 전용)
   */
  deleteUser = async (req: Request, res: Response): Promise<void> => {
    try {
      const id = parseInt(req.params.id);
      if (isNaN(id)) {
        res.status(400).json({
          error: '잘못된 사용자 ID',
          detail: '유효한 숫자 ID를 입력해주세요'
        });
        return;
      }

      await this.userService.deleteUser(id);

      const response: ApiResponse = {
        success: true,
        message: '사용자가 삭제되었습니다'
      };

      res.json(response);
    } catch (error) {
      res.status(400).json({
        error: '사용자 삭제 실패',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };

  /**
   * 사용자 검색
   */
  searchUsers = async (req: Request, res: Response): Promise<void> => {
    try {
      const name = req.query.name as string;
      if (!name) {
        res.status(400).json({
          error: '검색어 필요',
          detail: 'name 파라미터를 입력해주세요'
        });
        return;
      }

      const users = await this.userService.searchUsersByName(name);
      res.json(users);
    } catch (error) {
      res.status(500).json({
        error: '사용자 검색 실패',
        detail: error instanceof Error ? error.message : '알 수 없는 오류'
      });
    }
  };
} 

File: backend-java/src/main/java/com/example/intranet/controller/UserController.java
package com.example.intranet.controller;

import com.example.intranet.dto.*;
import com.example.intranet.entity.User;
import com.example.intranet.service.UserService;
import com.example.intranet.util.JwtUtil;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * 사용자 관리 REST API 컨트롤러
 * Go, Python 백엔드와 동일한 API 스펙 제공
 */
@RestController
@RequestMapping("/api/v1")
@CrossOrigin(origins = "*")
public class UserController {

    private final UserService userService;
    private final JwtUtil jwtUtil;

    @Autowired
    public UserController(UserService userService, JwtUtil jwtUtil) {
        this.userService = userService;
        this.jwtUtil = jwtUtil;
    }

    /**
     * 헬스체크 엔드포인트
     */
    @GetMapping("/health")
    public ResponseEntity<Map<String, Object>> healthCheck() {
        return ResponseEntity.ok(Map.of(
                "status", "healthy",
                "timestamp", System.currentTimeMillis() / 1000,
                "app_name", "Sample Intranet Backend Java"
        ));
    }

    /**
     * 회원가입
     */
    @PostMapping("/auth/register")
    public ResponseEntity<?> register(@Valid @RequestBody UserCreateRequest request) {
        try {
            UserResponse userResponse = userService.createUser(request);
            return ResponseEntity.status(HttpStatus.CREATED).body(userResponse);
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "회원가입 실패",
                    "detail", e.getMessage()
            ));
        }
    }

    /**
     * 로그인
     */
    @PostMapping("/auth/login")
    public ResponseEntity<?> login(@Valid @RequestBody UserLoginRequest request) {
        try {
            Optional<User> userOpt = userService.authenticateUser(request.getEmail(), request.getPassword());
            
            if (userOpt.isPresent()) {
                User user = userOpt.get();
                String token = jwtUtil.generateToken(user.getEmail(), user.getId());
                
                JwtResponse response = new JwtResponse(
                        token,
                        jwtUtil.getExpirationInSeconds(),
                        UserResponse.from(user)
                );
                
                return ResponseEntity.ok(response);
            } else {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of(
                        "error", "로그인 실패",
                        "detail", "이메일 또는 비밀번호가 올바르지 않습니다"
                ));
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                    "error", "로그인 처리 중 오류 발생",
                    "detail", e.getMessage()
            ));
        }
    }

    /**
     * 사용자 목록 조회 (인증 필요)
     */
    @GetMapping("/users")
    public ResponseEntity<List<UserResponse>> getUsers(
            @RequestParam(defaultValue = "0") int skip,
            @RequestParam(defaultValue = "100") int limit) {
        // TODO: JWT 인증 미들웨어 추가 필요
        List<UserResponse> users = userService.getAllUsers();
        return ResponseEntity.ok(users);
    }

    /**
     * 특정 사용자 조회 (인증 필요)
     */
    @GetMapping("/users/{id}")
    public ResponseEntity<?> getUser(@PathVariable Long id) {
        try {
            UserResponse user = userService.getUserById(id);
            return ResponseEntity.ok(user);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    /**
     * 사용자 정보 수정 (인증 필요)
     */
    @PutMapping("/users/{id}")
    public ResponseEntity<?> updateUser(
            @PathVariable Long id,
            @RequestBody Map<String, Object> updates) {
        try {
            // TODO: 권한 체크 (본인 또는 관리자만)
            String name = (String) updates.get("name");
            String roleStr = (String) updates.get("role");
            Boolean isActive = (Boolean) updates.get("isActive");
            
            UserResponse updated = userService.updateUser(id, name, 
                    roleStr != null ? com.example.intranet.entity.UserRole.valueOf(roleStr.toUpperCase()) : null, 
                    isActive);
            
            return ResponseEntity.ok(updated);
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "사용자 정보 수정 실패",
                    "detail", e.getMessage()
            ));
        }
    }

    /**
     * 사용자 삭제 (관리자 전용)
     */
    @DeleteMapping("/admin/users/{id}")
    public ResponseEntity<?> deleteUser(@PathVariable Long id) {
        try {
            // TODO: 관리자 권한 체크
            userService.deleteUser(id);
            return ResponseEntity.ok(Map.of(
                    "message", "사용자가 삭제되었습니다"
            ));
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "사용자 삭제 실패",
                    "detail", e.getMessage()
            ));
        }
    }

    /**
     * 사용자 검색
     */
    @GetMapping("/users/search")
    public ResponseEntity<List<UserResponse>> searchUsers(@RequestParam String name) {
        List<UserResponse> users = userService.searchUsersByName(name);
        return ResponseEntity.ok(users);
    }
} 